### PR TITLE
ENG-12759: faster and better tested serialization implementation

### PR DESCRIPTION
**Pull request by gernest**
I have mainly focused on reducing allocations, the following is the allocation comparison against current master branch.

```
benchmark                                                       old allocs     new allocs     delta
BenchmarkSerializeArgs-4                                        50             8              -84.00%
BenchmarkDeserializeResponse-4                                  6              6              +0.00%
BenchmarkDeserializeRows-4                                      848            848            +0.00%
BenchmarkDeserializeResult-4                                    12             12             +0.00%
BenchmarkExec-4                                                 534            495            -7.30%
BenchmarkPrepareStmtExec-4                                      522            486            -6.90%
BenchmarkQuery-4                                                1837           1896           +3.21%
BenchmarkPrepareStmtQuery-4                                     2030           1890           -6.90%
BenchmarkHashinater_getHashedPartitionForParameter_int32-4      1              1              +0.00%
BenchmarkHashinater_getHashedPartitionForParameter_int64-4      1              1              +0.00%
BenchmarkHashinater_getHashedPartitionForParameter_String-4     1              1              +0.00%
BenchmarkHashinater_getHashedPartitionForParameter_Bytes-4      0              0              +0.00%
```

### CODE REVIEW

File: wire/encoder.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+		return e.MarshalInt32(x)
+	case int64:
+		return e.MarshalInt64(x)
+	case float64:
+		return e.MarshalFloat64(x)
+	case string:
+		return e.MarshalString(x)
+	case time.Time:
+		return e.MarshalTime(x)
+	default:
```
Comment (nshi) [inline wire/encoder.go]: What about Table and Decimal types?

File: wire/encoder.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	i, err := e.Bool(v)
+	if err != nil {
+		return 0, err
+	}
+	return n + i, nil
+}
+
+// MarshalByte encodes int8 argument
+func (e *Encoder) MarshalByte(v int8) (int, error) {
+	n, err := e.Byte(BoolColumn)
```
Comment (nshi) [inline wire/encoder.go]: It is a bit confusing here that the type is `BoolColumn`. Maybe it makes more sense to name it `ByteColumn`, meaning that a bool can be represented as a byte, not vice versa.

File: wire/encoder.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+		if err != nil {
+			return 0, err
+		}
+		return n + i, nil
+	default:
+		n, err := e.Byte(ArrayColumn)
+		if err != nil {
+			return 0, err
+		}
+		s, err := e.Int16(1)
```
Comment (nshi) [inline wire/encoder.go]: Shouldn't this be the length of the slice? Why is it 1?

File: voltdbclient/authentication_test.go [hunk 4]
```diff
@@ -23,21 +23,6 @@ import (
 	"testing"
 )
 
-func TestLoginRequest(t *testing.T) {
```
Comment (gernest) [inline voltdbclient/authentication_test.go]: @nshi my bad, the name of the test should have been `TestEncoder_Login` does this sound okay to you?

The answer to your question is no, this is the test for login encoded message

File: wire/encoder.go (hunk 1 above)
Comment (gernest) [inline wire/encoder.go]: My first task it to meke sure I don't break a thing. There is no decimal types in the client, please open a ticket for it if you want it supported. My impression is that can wait untill all the pieces are in place

File: wire/encoder.go (hunk 2 above)
Comment (gernest) [inline wire/encoder.go]: They are the same value, but I do agree. I will make an alias.

File: voltdbclient/authentication_test.go (hunk 4 above)
Comment (gernest) [inline voltdbclient/authentication_test.go]: I'm sorry, I answered this in a hurry. Yes this is tested in the `wire` subpackage

File: wire/encoder.go (hunk 1 above)
Comment (gernest) [inline wire/encoder.go]: I took a look into the wire protocol again. The `Encoder.Marshal` method only encodes parameters, there is no mention of `Table` type as parameter, unless if I missed something.

My understanding is, we receive tables from the database we don't send them.

**Pull request merged**